{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Notebook to generate mutation dataset for a single date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "from tqdm import tqdm\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from collections import Counter\n",
    "from datetime import datetime as dt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### JH Time Series Data\n",
    "\n",
    "* Time series data is cumlative per column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "date_up_to = '10/29/20'\n",
    "jh_path = '../COVID-19/csse_covid_19_data/csse_covid_19_time_series/'\n",
    "confirmed_global = 'time_series_covid19_confirmed_global.csv'\n",
    "deaths_global = 'time_series_covid19_deaths_global.csv'\n",
    "\n",
    "confirmed = pd.read_csv(jh_path+confirmed_global)\n",
    "deaths = pd.read_csv(jh_path+deaths_global)\n",
    "\n",
    "#Need to adjust country names to match those in CNCB\n",
    "confirmed.loc[confirmed['Country/Region']=='US','Country/Region'] = 'United States'\n",
    "confirmed.loc[confirmed['Country/Region']=='Congo (Kinshasa)','Country/Region'] = 'Democratic Republic of the Congo'\n",
    "confirmed.loc[confirmed['Country/Region']=='Korea, South','Country/Region'] = 'South Korea'\n",
    "confirmed.loc[confirmed['Country/Region']=='Czechia','Country/Region'] = 'Czech Republic'\n",
    "confirmed.loc[confirmed['Country/Region']=='Burma','Country/Region'] = 'myanmar'\n",
    "confirmed.loc[confirmed['Country/Region']=='Congo (Brazzaville)','Country/Region'] = 'republic of the congo'\n",
    "\n",
    "confirmed['Country/Region'] = confirmed['Country/Region'].str.lower()\n",
    "\n",
    "deaths.loc[deaths['Country/Region']=='US','Country/Region'] = 'United States'\n",
    "deaths.loc[deaths['Country/Region']=='Congo (Kinshasa)','Country/Region'] = 'Democratic Republic of the Congo'\n",
    "deaths.loc[deaths['Country/Region']=='Korea, South','Country/Region'] = 'South Korea'\n",
    "deaths.loc[deaths['Country/Region']=='Czechia','Country/Region'] = 'Czech Republic'\n",
    "deaths.loc[deaths['Country/Region']=='Burma','Country/Region'] = 'myanmar'\n",
    "deaths.loc[deaths['Country/Region']=='Congo (Brazzaville)','Country/Region'] = 'republic of the congo'\n",
    "\n",
    "deaths['Country/Region'] = deaths['Country/Region'].str.lower()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Need other data table for population number\n",
    "* doesn't matter for date as long as all the countries are covered, only getting population number"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "date = '10-29-2020.csv'\n",
    "jh_path = '../COVID-19/csse_covid_19_data/csse_covid_19_daily_reports/'\n",
    "jh_data = pd.read_csv(jh_path+date)\n",
    "\n",
    "#drop rows that have nans for incidence rate\n",
    "jh_data = jh_data[~jh_data['Incidence_Rate'].isnull()]\n",
    "jh_data = jh_data[jh_data['Incidence_Rate']!=0]\n",
    "\n",
    "#matching countries of jh data to that of cncb countries\n",
    "jh_data.loc[jh_data['Country_Region']=='Taiwan*','Country_Region'] = 'Taiwan'\n",
    "jh_data.loc[jh_data['Country_Region']=='US','Country_Region'] = 'United States'\n",
    "jh_data.loc[jh_data['Country_Region']=='Korea, South','Country_Region'] = 'South Korea'\n",
    "jh_data.loc[jh_data['Country_Region']=='Czechia','Country_Region'] = 'Czech Republic'\n",
    "jh_data.loc[jh_data['Country_Region']=='Burma','Country_Region'] = 'myanmar'\n",
    "jh_data.loc[jh_data['Country_Region']=='Congo (Kinshasa)','Country_Region'] = 'Democratic Republic of the Congo'\n",
    "jh_data.loc[jh_data['Country_Region']=='Congo (Brazzaville)','Country_Region'] = 'republic of the congo'\n",
    "\n",
    "jh_data['Population'] = jh_data['Confirmed']/jh_data['Incidence_Rate']*100000\n",
    "jh_data['Country_Region'] = jh_data['Country_Region'].str.lower()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CNCB Genome Meta Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(93012, 9)\n"
     ]
    }
   ],
   "source": [
    "#cncb meta, need to match countries\n",
    "column_names = ['Virus Strain Name','Accession ID','Related ID','Nuc.Completeness','Sequence Quality','Host','Location','Sample Collection Date','Submitting Lab']\n",
    "meta = pd.read_excel('../metadata10_27.xlsx',usecols=column_names,index_col=1)\n",
    "meta.fillna(' ',inplace=True)\n",
    "meta['Country'] = meta['Location'].str.split('/').str[0].str.strip()\n",
    "meta.loc[meta['Country']=='\\u200eRomania','Country'] = 'Romania'\n",
    "#Remove low quality and partial reads, pretty sure cncb does not run variant annotation for these anyways\n",
    "meta = meta[meta['Sequence Quality']!='Low']\n",
    "meta = meta[meta['Nuc.Completeness']!='Partial']\n",
    "meta['Country'] = meta['Country'].str.lower()\n",
    "\n",
    "#Filter using date as well\n",
    "meta = meta[meta['Sample Collection Date']!='2020-00-00'] #bad dates in there\n",
    "meta.loc[:,'Sample Collection Date'] = pd.to_datetime(meta['Sample Collection Date'], yearfirst=True)\n",
    "dt_date_up_to = dt.strptime(date_up_to, \"%m/%d/%y\")\n",
    "meta = meta[meta.loc[:,'Sample Collection Date']<dt_date_up_to]\n",
    "\n",
    "print(meta.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### If there is a missing country in set difference, there will be nan data in output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(meta['Country'])-set(confirmed['Country/Region'].str.lower())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "110"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(meta['Country']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Using Infections per 100k and Fatality percentage\n",
    "* Currently summing over country labels, may be incorrect for some with territories, minimal impact"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(110, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cases</th>\n",
       "      <th>deaths</th>\n",
       "      <th>population</th>\n",
       "      <th>Fatality_rate</th>\n",
       "      <th>Incidence_rate_per100k</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>algeria</th>\n",
       "      <td>57332.0</td>\n",
       "      <td>1949.0</td>\n",
       "      <td>43851043.0</td>\n",
       "      <td>3.399498</td>\n",
       "      <td>130.742614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>andorra</th>\n",
       "      <td>4567.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>77265.0</td>\n",
       "      <td>1.598423</td>\n",
       "      <td>5910.826377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>argentina</th>\n",
       "      <td>1143800.0</td>\n",
       "      <td>30442.0</td>\n",
       "      <td>45195777.0</td>\n",
       "      <td>2.661479</td>\n",
       "      <td>2530.767421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>australia</th>\n",
       "      <td>27579.0</td>\n",
       "      <td>907.0</td>\n",
       "      <td>25459700.0</td>\n",
       "      <td>3.288734</td>\n",
       "      <td>108.324136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>austria</th>\n",
       "      <td>93949.0</td>\n",
       "      <td>1056.0</td>\n",
       "      <td>9006400.0</td>\n",
       "      <td>1.124014</td>\n",
       "      <td>1043.135992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>united states</th>\n",
       "      <td>8944934.0</td>\n",
       "      <td>228656.0</td>\n",
       "      <td>332804239.0</td>\n",
       "      <td>2.556263</td>\n",
       "      <td>2687.746414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>uruguay</th>\n",
       "      <td>3044.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>3473727.0</td>\n",
       "      <td>1.872536</td>\n",
       "      <td>87.629224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>venezuela</th>\n",
       "      <td>91280.0</td>\n",
       "      <td>789.0</td>\n",
       "      <td>28435943.0</td>\n",
       "      <td>0.864373</td>\n",
       "      <td>321.002191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vietnam</th>\n",
       "      <td>1177.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>97338583.0</td>\n",
       "      <td>2.973662</td>\n",
       "      <td>1.209181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zambia</th>\n",
       "      <td>16325.0</td>\n",
       "      <td>348.0</td>\n",
       "      <td>18383956.0</td>\n",
       "      <td>2.131700</td>\n",
       "      <td>88.800256</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>110 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   cases    deaths   population  Fatality_rate  \\\n",
       "algeria          57332.0    1949.0   43851043.0       3.399498   \n",
       "andorra           4567.0      73.0      77265.0       1.598423   \n",
       "argentina      1143800.0   30442.0   45195777.0       2.661479   \n",
       "australia        27579.0     907.0   25459700.0       3.288734   \n",
       "austria          93949.0    1056.0    9006400.0       1.124014   \n",
       "...                  ...       ...          ...            ...   \n",
       "united states  8944934.0  228656.0  332804239.0       2.556263   \n",
       "uruguay           3044.0      57.0    3473727.0       1.872536   \n",
       "venezuela        91280.0     789.0   28435943.0       0.864373   \n",
       "vietnam           1177.0      35.0   97338583.0       2.973662   \n",
       "zambia           16325.0     348.0   18383956.0       2.131700   \n",
       "\n",
       "               Incidence_rate_per100k  \n",
       "algeria                    130.742614  \n",
       "andorra                   5910.826377  \n",
       "argentina                 2530.767421  \n",
       "australia                  108.324136  \n",
       "austria                   1043.135992  \n",
       "...                               ...  \n",
       "united states             2687.746414  \n",
       "uruguay                     87.629224  \n",
       "venezuela                  321.002191  \n",
       "vietnam                      1.209181  \n",
       "zambia                      88.800256  \n",
       "\n",
       "[110 rows x 5 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Using coutries from meta, sum values to get counts per country\n",
    "jh_country_lvl = pd.DataFrame(columns=['cases','deaths','population','Fatality_rate','Incidence_rate_per100k'])\n",
    "for country in set(meta['Country'].str.lower()):\n",
    "    summed_cases = confirmed.loc[confirmed['Country/Region']==country,date_up_to].sum()\n",
    "    summed_deaths = deaths.loc[deaths['Country/Region']==country,date_up_to].sum()\n",
    "    population = jh_data.loc[jh_data['Country_Region']==country,'Population'].sum()\n",
    "    if summed_cases == 0:\n",
    "        fate_rate = 0\n",
    "    else:\n",
    "        fate_rate = 100*summed_deaths/summed_cases\n",
    "    inc_rate = 100000*summed_cases/population\n",
    "    jh_country_lvl.loc[country] = [summed_cases,summed_deaths,population,fate_rate,inc_rate]\n",
    "        \n",
    "    \n",
    "# jh_country_lvl['Fatality_rate'] = 100*jh_country_lvl['Deaths']/jh_country_lvl['Confirmed']\n",
    "# jh_country_lvl['Incidence_rate_per100k'] = 100000*jh_country_lvl['Confirmed']/jh_country_lvl['Population']\n",
    "# jh_country_lvl.replace([np.inf, -np.inf], 0,inplace=True)\n",
    "print(jh_country_lvl.shape)\n",
    "jh_country_lvl.sort_index(inplace=True)\n",
    "jh_country_lvl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parsing gff3 files\n",
    "* Currently parse everytime you need to change the date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_names = os.listdir('../gff3_cncb')\n",
    "column_names = ['variant type','start','end', 'info']\n",
    "missing = []\n",
    "aggregated_mutations = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 67766/67766 [32:24<00:00, 34.85it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "all_names = os.listdir('../gff3_cncb_restructured')\n",
    "column_names = ['variant type','start','end', 'info']\n",
    "missing = []\n",
    "aggregated_mutations = {}\n",
    "\n",
    "#for identifier in tqdm(meta[meta['Country']=='United States'].index):\n",
    "for identifier in tqdm(meta.index):\n",
    "    \n",
    "    #Searching for correct identifier\n",
    "    #--------------------------\n",
    "    #No alternate name is ' '\n",
    "    file_name = ''\n",
    "    #Check if accession id in file names, if not check related ids\n",
    "    if '2019-nCoV_'+identifier+'_variants.gff3' in all_names:\n",
    "        file_name = '2019-nCoV_'+identifier+'_variants.gff3'\n",
    "    # checking alternate names\n",
    "    elif meta.loc[identifier,'Related ID'] != ' ':\n",
    "        for alt_identifier in meta.loc[identifier,'Related ID'].replace(' ','').split(','):\n",
    "            if '2019-nCoV_'+alt_identifier+'_variants.gff3' in all_names:\n",
    "                file_name = '2019-nCoV_'+alt_identifier+'_variants.gff3'\n",
    "                break\n",
    "        #Added in case alternate names are also not found in gffs\n",
    "        if file_name == '':\n",
    "            missing.append(identifier)\n",
    "            continue\n",
    "    # If file name has not been updated, then there is no matching identifier, move to next index\n",
    "    elif file_name == '':\n",
    "        missing.append(identifier)\n",
    "        continue\n",
    "    #--------------------------\n",
    "    \n",
    "    #Filtering files with no variants\n",
    "    #--------------------------\n",
    "    with open(f'../gff3_cncb_restructured/{file_name}') as text_file:\n",
    "        lines = text_file.readlines()\n",
    "        counter = 0\n",
    "        for l in lines:\n",
    "            if '#' in l:\n",
    "                counter += 1\n",
    "    #Number of info lines should be less than total, if not then there are no mutations\n",
    "    #--------------------------\n",
    "    if counter<len(lines) and meta.loc[identifier,'Country'] in set(jh_data['Country_Region']): #country needs to be found in jh data for inf/fata rates\n",
    "        \n",
    "        gff = pd.read_csv(f'../gff3_cncb_restructured/{file_name}',sep='\\t',skiprows=counter,usecols=[1,3,4,8],names=column_names)\n",
    "        info_df = pd.DataFrame(gff['info'].str.split(';').values.tolist(),columns=[0,1,'Ref','Alt','Description']).drop([0,1],axis=1)\n",
    "        gff = gff.drop(['info'],axis=1)\n",
    "        gff['Country'] = [meta.loc[identifier,'Country']]*gff.shape[0]\n",
    "        gff['infection_rate'] = [jh_country_lvl.loc[meta.loc[identifier,'Country'],'Incidence_rate_per100k']]*gff.shape[0]\n",
    "        gff['fatality_rate'] = [jh_country_lvl.loc[meta.loc[identifier,'Country'],'Fatality_rate']]*gff.shape[0]\n",
    "        temp_df = pd.concat([gff,info_df],axis=1)\n",
    "        \n",
    "        #Filtering alternate amino acid and reference for missense_variant and synonymous_variant\n",
    "        missenses_ref = temp_df.loc[temp_df['Description'].str.contains('missense_variant'),'Description'].str.split(',').str[1].str[-3]\n",
    "        synonymous_ref = temp_df.loc[temp_df['Description'].str.contains('synonymous_variant'),'Description'].str.split(',').str[1].str[-1]\n",
    "        temp_df['Ref_AA'] = pd.concat([missenses_ref,synonymous_ref])\n",
    "        temp_df['Alt_AA'] = temp_df.loc[temp_df['Description'].str.contains('missense_variant'),'Description'].str.split(',').str[1].str[-1]\n",
    "\n",
    "        missenses_str = temp_df.loc[temp_df['Description'].str.contains('missense_variant'),'Description'].str.split(',').str[1].str.split('.').str[-1]\n",
    "        synonymous_str = temp_df.loc[temp_df['Description'].str.contains('synonymous_variant'),'Description'].str.split(',').str[1].str.split('.').str[-1]\n",
    "        temp_df['AA'] = pd.concat([missenses_str,synonymous_str])\n",
    "        \n",
    "        temp_df.fillna('', inplace=True)\n",
    "        temp_df['descriptor'] = temp_df['start'].astype(str)+','+temp_df['end'].astype(str)+','+temp_df['Ref']+','+temp_df['Alt']+\\\n",
    "        ','+temp_df['Description'].str.split(',').str[0].str.split('=').str[1]+','+temp_df['variant type']+','+temp_df['Ref_AA']+','+temp_df['Alt_AA']+\\\n",
    "        ','+temp_df['AA']\n",
    "\n",
    "#         if any(temp_df['Ref'].str.contains('REF=A') & temp_df['Alt'].str.contains('ALT=GGTTC')):\n",
    "#             break\n",
    "        \n",
    "        for var in temp_df['descriptor']:\n",
    "            if var not in aggregated_mutations.keys():\n",
    "                aggregated_mutations[var] = [[],[],[]]\n",
    "                \n",
    "            #Below numbers should be the same for each mutation of same file\n",
    "            #Country\n",
    "            aggregated_mutations[var][0].append(temp_df.loc[0,'Country'])\n",
    "            #Infection Rate\n",
    "            aggregated_mutations[var][1].append(temp_df.loc[0,'infection_rate'])\n",
    "            #Fatality Rate\n",
    "            aggregated_mutations[var][2].append(temp_df.loc[0,'fatality_rate'])\n",
    "\n",
    "print(len(missing))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_muts = pd.DataFrame(list(aggregated_mutations.keys()),columns=['descriptor'])\n",
    "unique_muts = pd.DataFrame.join(unique_muts,pd.DataFrame(unique_muts['descriptor'].str.split(',').to_list())) #assign columns with parsed descriptor\n",
    "unique_muts.set_index('descriptor',inplace=True)\n",
    "unique_muts[0] = pd.to_numeric(unique_muts[0])\n",
    "unique_muts[1] = pd.to_numeric(unique_muts[1])\n",
    "unique_muts.sort_values([0,1],inplace=True)\n",
    "unique_muts.columns = ['Start','End','Ref','Alt','VEP','Variant Type','Ref_AA','Alt_AA','AA']\n",
    "\n",
    "\n",
    "counts = {}\n",
    "num_countries = {}\n",
    "infection_rate = {}\n",
    "fatality_rate = {}\n",
    "counted_countries = {}\n",
    "\n",
    "for desc in unique_muts.index:\n",
    "    counts[desc] = len(aggregated_mutations[desc][0])\n",
    "    num_countries[desc] = len(set(aggregated_mutations[desc][0]))\n",
    "    infection_rate[desc] = np.mean(aggregated_mutations[desc][1])\n",
    "    fatality_rate[desc] = np.mean(aggregated_mutations[desc][2])\n",
    "    counted_countries[desc] = dict(Counter(aggregated_mutations[desc][0]))\n",
    "        \n",
    "unique_muts['counts'] = unique_muts.index.to_series().map(counts)\n",
    "unique_muts['countries'] = unique_muts.index.to_series().map(num_countries)\n",
    "unique_muts['infection_rate'] = unique_muts.index.to_series().map(infection_rate)\n",
    "unique_muts['fatality_rate'] = unique_muts.index.to_series().map(fatality_rate)\n",
    "unique_muts['counted_countries'] = unique_muts.index.to_series().map(counted_countries)\n",
    "unique_muts.index = unique_muts.index.str.split(',').str[0:4].str.join('_')\n",
    "unique_muts.to_csv(f\"temporal_files/added_b/unique_mutations_country{'_'.join(date_up_to.split('/'))}.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>VEP</th>\n",
       "      <th>Variant Type</th>\n",
       "      <th>Ref_AA</th>\n",
       "      <th>Alt_AA</th>\n",
       "      <th>AA</th>\n",
       "      <th>counts</th>\n",
       "      <th>countries</th>\n",
       "      <th>infection_rate</th>\n",
       "      <th>fatality_rate</th>\n",
       "      <th>counted_countries</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>descriptor</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_1_REF=AA_ALT=-</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>REF=AA</td>\n",
       "      <td>ALT=-</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>Deletion</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>966</td>\n",
       "      <td>33</td>\n",
       "      <td>815.093885</td>\n",
       "      <td>4.996641</td>\n",
       "      <td>{'china': 15, 'singapore': 8, 'france': 2, 'ca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_1_REF=A_ALT=CCAGAAGAA</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>REF=A</td>\n",
       "      <td>ALT=CCAGAAGAA</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>Indel</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>6.136787</td>\n",
       "      <td>5.394306</td>\n",
       "      <td>{'china': 2}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_1_REF=A_ALT=G</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>REF=A</td>\n",
       "      <td>ALT=G</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>70</td>\n",
       "      <td>10</td>\n",
       "      <td>1064.553193</td>\n",
       "      <td>3.886944</td>\n",
       "      <td>{'china': 10, 'united states': 42, 'chile': 7,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_1_REF=A_ALT=CCCTGAATTAGACTCA</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>REF=A</td>\n",
       "      <td>ALT=CCCTGAATTAGACTCA</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>Indel</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>6.136787</td>\n",
       "      <td>5.394306</td>\n",
       "      <td>{'china': 1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_1_REF=A_ALT=GGTTC</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>REF=A</td>\n",
       "      <td>ALT=GGTTC</td>\n",
       "      <td>intergenic_variant||intergenic_variant</td>\n",
       "      <td>Indel</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>6.136787</td>\n",
       "      <td>5.394306</td>\n",
       "      <td>{'china': 1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29903_29903_REF=-_ALT=-ANANANAC</th>\n",
       "      <td>29903</td>\n",
       "      <td>29903</td>\n",
       "      <td>REF=-</td>\n",
       "      <td>ALT=-ANANANAC</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>Insertion</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1235.078018</td>\n",
       "      <td>3.558906</td>\n",
       "      <td>{'united states': 1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29903_29903_REF=-_ALT=-AAANANANANAAAAAAA</th>\n",
       "      <td>29903</td>\n",
       "      <td>29903</td>\n",
       "      <td>REF=-</td>\n",
       "      <td>ALT=-AAANANANANAAAAAAA</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>Insertion</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1235.078018</td>\n",
       "      <td>3.558906</td>\n",
       "      <td>{'united states': 1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29903_29903_REF=A_ALT=ACAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA</th>\n",
       "      <td>29903</td>\n",
       "      <td>29903</td>\n",
       "      <td>REF=A</td>\n",
       "      <td>ALT=ACAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>Insertion</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>580.396386</td>\n",
       "      <td>10.436787</td>\n",
       "      <td>{'spain': 1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29903_29903_REF=A_ALT=CTTTATTATTATGTAT</th>\n",
       "      <td>29903</td>\n",
       "      <td>29903</td>\n",
       "      <td>REF=A</td>\n",
       "      <td>ALT=CTTTATTATTATGTAT</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>Indel</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>260.329398</td>\n",
       "      <td>4.022500</td>\n",
       "      <td>{'iraq': 1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29903_29903_REF=A_ALT=AAAAAAAAAATTAAGAAAAAA</th>\n",
       "      <td>29903</td>\n",
       "      <td>29903</td>\n",
       "      <td>REF=A</td>\n",
       "      <td>ALT=AAAAAAAAAATTAAGAAAAAA</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>Insertion</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>687.132093</td>\n",
       "      <td>0.753258</td>\n",
       "      <td>{'israel': 1}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34220 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    Start    End     Ref  \\\n",
       "descriptor                                                                 \n",
       "1_1_REF=AA_ALT=-                                        1      1  REF=AA   \n",
       "1_1_REF=A_ALT=CCAGAAGAA                                 1      1   REF=A   \n",
       "1_1_REF=A_ALT=G                                         1      1   REF=A   \n",
       "1_1_REF=A_ALT=CCCTGAATTAGACTCA                          1      1   REF=A   \n",
       "1_1_REF=A_ALT=GGTTC                                     1      1   REF=A   \n",
       "...                                                   ...    ...     ...   \n",
       "29903_29903_REF=-_ALT=-ANANANAC                     29903  29903   REF=-   \n",
       "29903_29903_REF=-_ALT=-AAANANANANAAAAAAA            29903  29903   REF=-   \n",
       "29903_29903_REF=A_ALT=ACAAAAAAAAAAAAAAAAAAAAAAA...  29903  29903   REF=A   \n",
       "29903_29903_REF=A_ALT=CTTTATTATTATGTAT              29903  29903   REF=A   \n",
       "29903_29903_REF=A_ALT=AAAAAAAAAATTAAGAAAAAA         29903  29903   REF=A   \n",
       "\n",
       "                                                                                        Alt  \\\n",
       "descriptor                                                                                    \n",
       "1_1_REF=AA_ALT=-                                                                      ALT=-   \n",
       "1_1_REF=A_ALT=CCAGAAGAA                                                       ALT=CCAGAAGAA   \n",
       "1_1_REF=A_ALT=G                                                                       ALT=G   \n",
       "1_1_REF=A_ALT=CCCTGAATTAGACTCA                                         ALT=CCCTGAATTAGACTCA   \n",
       "1_1_REF=A_ALT=GGTTC                                                               ALT=GGTTC   \n",
       "...                                                                                     ...   \n",
       "29903_29903_REF=-_ALT=-ANANANAC                                               ALT=-ANANANAC   \n",
       "29903_29903_REF=-_ALT=-AAANANANANAAAAAAA                             ALT=-AAANANANANAAAAAAA   \n",
       "29903_29903_REF=A_ALT=ACAAAAAAAAAAAAAAAAAAAAAAA...  ALT=ACAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA   \n",
       "29903_29903_REF=A_ALT=CTTTATTATTATGTAT                                 ALT=CTTTATTATTATGTAT   \n",
       "29903_29903_REF=A_ALT=AAAAAAAAAATTAAGAAAAAA                       ALT=AAAAAAAAAATTAAGAAAAAA   \n",
       "\n",
       "                                                                                       VEP  \\\n",
       "descriptor                                                                                   \n",
       "1_1_REF=AA_ALT=-                                                        intergenic_variant   \n",
       "1_1_REF=A_ALT=CCAGAAGAA                                                 intergenic_variant   \n",
       "1_1_REF=A_ALT=G                                                         intergenic_variant   \n",
       "1_1_REF=A_ALT=CCCTGAATTAGACTCA                                          intergenic_variant   \n",
       "1_1_REF=A_ALT=GGTTC                                 intergenic_variant||intergenic_variant   \n",
       "...                                                                                    ...   \n",
       "29903_29903_REF=-_ALT=-ANANANAC                                         intergenic_variant   \n",
       "29903_29903_REF=-_ALT=-AAANANANANAAAAAAA                                intergenic_variant   \n",
       "29903_29903_REF=A_ALT=ACAAAAAAAAAAAAAAAAAAAAAAA...                      intergenic_variant   \n",
       "29903_29903_REF=A_ALT=CTTTATTATTATGTAT                                  intergenic_variant   \n",
       "29903_29903_REF=A_ALT=AAAAAAAAAATTAAGAAAAAA                             intergenic_variant   \n",
       "\n",
       "                                                   Variant Type Ref_AA Alt_AA  \\\n",
       "descriptor                                                                      \n",
       "1_1_REF=AA_ALT=-                                       Deletion                 \n",
       "1_1_REF=A_ALT=CCAGAAGAA                                   Indel                 \n",
       "1_1_REF=A_ALT=G                                             SNP                 \n",
       "1_1_REF=A_ALT=CCCTGAATTAGACTCA                            Indel                 \n",
       "1_1_REF=A_ALT=GGTTC                                       Indel                 \n",
       "...                                                         ...    ...    ...   \n",
       "29903_29903_REF=-_ALT=-ANANANAC                       Insertion                 \n",
       "29903_29903_REF=-_ALT=-AAANANANANAAAAAAA              Insertion                 \n",
       "29903_29903_REF=A_ALT=ACAAAAAAAAAAAAAAAAAAAAAAA...    Insertion                 \n",
       "29903_29903_REF=A_ALT=CTTTATTATTATGTAT                    Indel                 \n",
       "29903_29903_REF=A_ALT=AAAAAAAAAATTAAGAAAAAA           Insertion                 \n",
       "\n",
       "                                                   AA  counts  countries  \\\n",
       "descriptor                                                                 \n",
       "1_1_REF=AA_ALT=-                                          966         33   \n",
       "1_1_REF=A_ALT=CCAGAAGAA                                     2          1   \n",
       "1_1_REF=A_ALT=G                                            70         10   \n",
       "1_1_REF=A_ALT=CCCTGAATTAGACTCA                              1          1   \n",
       "1_1_REF=A_ALT=GGTTC                                         1          1   \n",
       "...                                                ..     ...        ...   \n",
       "29903_29903_REF=-_ALT=-ANANANAC                             1          1   \n",
       "29903_29903_REF=-_ALT=-AAANANANANAAAAAAA                    1          1   \n",
       "29903_29903_REF=A_ALT=ACAAAAAAAAAAAAAAAAAAAAAAA...          1          1   \n",
       "29903_29903_REF=A_ALT=CTTTATTATTATGTAT                      1          1   \n",
       "29903_29903_REF=A_ALT=AAAAAAAAAATTAAGAAAAAA                 1          1   \n",
       "\n",
       "                                                    infection_rate  \\\n",
       "descriptor                                                           \n",
       "1_1_REF=AA_ALT=-                                        815.093885   \n",
       "1_1_REF=A_ALT=CCAGAAGAA                                   6.136787   \n",
       "1_1_REF=A_ALT=G                                        1064.553193   \n",
       "1_1_REF=A_ALT=CCCTGAATTAGACTCA                            6.136787   \n",
       "1_1_REF=A_ALT=GGTTC                                       6.136787   \n",
       "...                                                            ...   \n",
       "29903_29903_REF=-_ALT=-ANANANAC                        1235.078018   \n",
       "29903_29903_REF=-_ALT=-AAANANANANAAAAAAA               1235.078018   \n",
       "29903_29903_REF=A_ALT=ACAAAAAAAAAAAAAAAAAAAAAAA...      580.396386   \n",
       "29903_29903_REF=A_ALT=CTTTATTATTATGTAT                  260.329398   \n",
       "29903_29903_REF=A_ALT=AAAAAAAAAATTAAGAAAAAA             687.132093   \n",
       "\n",
       "                                                    fatality_rate  \\\n",
       "descriptor                                                          \n",
       "1_1_REF=AA_ALT=-                                         4.996641   \n",
       "1_1_REF=A_ALT=CCAGAAGAA                                  5.394306   \n",
       "1_1_REF=A_ALT=G                                          3.886944   \n",
       "1_1_REF=A_ALT=CCCTGAATTAGACTCA                           5.394306   \n",
       "1_1_REF=A_ALT=GGTTC                                      5.394306   \n",
       "...                                                           ...   \n",
       "29903_29903_REF=-_ALT=-ANANANAC                          3.558906   \n",
       "29903_29903_REF=-_ALT=-AAANANANANAAAAAAA                 3.558906   \n",
       "29903_29903_REF=A_ALT=ACAAAAAAAAAAAAAAAAAAAAAAA...      10.436787   \n",
       "29903_29903_REF=A_ALT=CTTTATTATTATGTAT                   4.022500   \n",
       "29903_29903_REF=A_ALT=AAAAAAAAAATTAAGAAAAAA              0.753258   \n",
       "\n",
       "                                                                                    counted_countries  \n",
       "descriptor                                                                                             \n",
       "1_1_REF=AA_ALT=-                                    {'china': 15, 'singapore': 8, 'france': 2, 'ca...  \n",
       "1_1_REF=A_ALT=CCAGAAGAA                                                                  {'china': 2}  \n",
       "1_1_REF=A_ALT=G                                     {'china': 10, 'united states': 42, 'chile': 7,...  \n",
       "1_1_REF=A_ALT=CCCTGAATTAGACTCA                                                           {'china': 1}  \n",
       "1_1_REF=A_ALT=GGTTC                                                                      {'china': 1}  \n",
       "...                                                                                               ...  \n",
       "29903_29903_REF=-_ALT=-ANANANAC                                                  {'united states': 1}  \n",
       "29903_29903_REF=-_ALT=-AAANANANANAAAAAAA                                         {'united states': 1}  \n",
       "29903_29903_REF=A_ALT=ACAAAAAAAAAAAAAAAAAAAAAAA...                                       {'spain': 1}  \n",
       "29903_29903_REF=A_ALT=CTTTATTATTATGTAT                                                    {'iraq': 1}  \n",
       "29903_29903_REF=A_ALT=AAAAAAAAAATTAAGAAAAAA                                             {'israel': 1}  \n",
       "\n",
       "[34220 rows x 14 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_muts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>VEP</th>\n",
       "      <th>Variant Type</th>\n",
       "      <th>Ref_AA</th>\n",
       "      <th>Alt_AA</th>\n",
       "      <th>AA</th>\n",
       "      <th>counts</th>\n",
       "      <th>countries</th>\n",
       "      <th>infection_rate</th>\n",
       "      <th>fatality_rate</th>\n",
       "      <th>counted_countries</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>descriptor</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>269_269_REF=G_ALT=A</th>\n",
       "      <td>269</td>\n",
       "      <td>269</td>\n",
       "      <td>REF=G</td>\n",
       "      <td>ALT=A</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>E</td>\n",
       "      <td>K</td>\n",
       "      <td>2E&gt;K</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>447.212807</td>\n",
       "      <td>13.747245</td>\n",
       "      <td>{'united kingdom': 1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270_270_REF=A_ALT=G</th>\n",
       "      <td>270</td>\n",
       "      <td>270</td>\n",
       "      <td>REF=A</td>\n",
       "      <td>ALT=G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>E</td>\n",
       "      <td>G</td>\n",
       "      <td>2E&gt;G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>6.136787</td>\n",
       "      <td>5.394306</td>\n",
       "      <td>{'china': 1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>271_271_REF=G_ALT=T</th>\n",
       "      <td>271</td>\n",
       "      <td>271</td>\n",
       "      <td>REF=G</td>\n",
       "      <td>ALT=T</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>E</td>\n",
       "      <td>D</td>\n",
       "      <td>2E&gt;D</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>447.212807</td>\n",
       "      <td>13.747245</td>\n",
       "      <td>{'united kingdom': 3}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>273_273_REF=G_ALT=A</th>\n",
       "      <td>273</td>\n",
       "      <td>273</td>\n",
       "      <td>REF=G</td>\n",
       "      <td>ALT=A</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>S</td>\n",
       "      <td>N</td>\n",
       "      <td>3S&gt;N</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1235.078018</td>\n",
       "      <td>3.558906</td>\n",
       "      <td>{'united states': 1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274_274_REF=C_ALT=A</th>\n",
       "      <td>274</td>\n",
       "      <td>274</td>\n",
       "      <td>REF=C</td>\n",
       "      <td>ALT=A</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>S</td>\n",
       "      <td>R</td>\n",
       "      <td>3S&gt;R</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1235.078018</td>\n",
       "      <td>3.558906</td>\n",
       "      <td>{'united states': 2}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29666_29666_REF=C_ALT=A</th>\n",
       "      <td>29666</td>\n",
       "      <td>29666</td>\n",
       "      <td>REF=C</td>\n",
       "      <td>ALT=A</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>L</td>\n",
       "      <td>I</td>\n",
       "      <td>37L&gt;I</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>447.212807</td>\n",
       "      <td>13.747245</td>\n",
       "      <td>{'united kingdom': 1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29666_29666_REF=C_ALT=G</th>\n",
       "      <td>29666</td>\n",
       "      <td>29666</td>\n",
       "      <td>REF=C</td>\n",
       "      <td>ALT=G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>L</td>\n",
       "      <td>V</td>\n",
       "      <td>37L&gt;V</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>447.212807</td>\n",
       "      <td>13.747245</td>\n",
       "      <td>{'united kingdom': 1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29667_29667_REF=T_ALT=C</th>\n",
       "      <td>29667</td>\n",
       "      <td>29667</td>\n",
       "      <td>REF=T</td>\n",
       "      <td>ALT=C</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>L</td>\n",
       "      <td>P</td>\n",
       "      <td>37L&gt;P</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1235.078018</td>\n",
       "      <td>3.558906</td>\n",
       "      <td>{'united states': 1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29669_29669_REF=A_ALT=G</th>\n",
       "      <td>29669</td>\n",
       "      <td>29669</td>\n",
       "      <td>REF=A</td>\n",
       "      <td>ALT=G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>38T&gt;A</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>22.892683</td>\n",
       "      <td>3.441471</td>\n",
       "      <td>{'japan': 1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29670_29670_REF=C_ALT=T</th>\n",
       "      <td>29670</td>\n",
       "      <td>29670</td>\n",
       "      <td>REF=C</td>\n",
       "      <td>ALT=T</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>I</td>\n",
       "      <td>38T&gt;I</td>\n",
       "      <td>13</td>\n",
       "      <td>7</td>\n",
       "      <td>590.213587</td>\n",
       "      <td>6.685468</td>\n",
       "      <td>{'china': 1, 'australia': 1, 'united kingdom':...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13244 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         Start    End    Ref    Alt               VEP  \\\n",
       "descriptor                                                              \n",
       "269_269_REF=G_ALT=A        269    269  REF=G  ALT=A  missense_variant   \n",
       "270_270_REF=A_ALT=G        270    270  REF=A  ALT=G  missense_variant   \n",
       "271_271_REF=G_ALT=T        271    271  REF=G  ALT=T  missense_variant   \n",
       "273_273_REF=G_ALT=A        273    273  REF=G  ALT=A  missense_variant   \n",
       "274_274_REF=C_ALT=A        274    274  REF=C  ALT=A  missense_variant   \n",
       "...                        ...    ...    ...    ...               ...   \n",
       "29666_29666_REF=C_ALT=A  29666  29666  REF=C  ALT=A  missense_variant   \n",
       "29666_29666_REF=C_ALT=G  29666  29666  REF=C  ALT=G  missense_variant   \n",
       "29667_29667_REF=T_ALT=C  29667  29667  REF=T  ALT=C  missense_variant   \n",
       "29669_29669_REF=A_ALT=G  29669  29669  REF=A  ALT=G  missense_variant   \n",
       "29670_29670_REF=C_ALT=T  29670  29670  REF=C  ALT=T  missense_variant   \n",
       "\n",
       "                        Variant Type Ref_AA Alt_AA     AA  counts  countries  \\\n",
       "descriptor                                                                     \n",
       "269_269_REF=G_ALT=A              SNP      E      K   2E>K       1          1   \n",
       "270_270_REF=A_ALT=G              SNP      E      G   2E>G       1          1   \n",
       "271_271_REF=G_ALT=T              SNP      E      D   2E>D       3          1   \n",
       "273_273_REF=G_ALT=A              SNP      S      N   3S>N       1          1   \n",
       "274_274_REF=C_ALT=A              SNP      S      R   3S>R       2          1   \n",
       "...                              ...    ...    ...    ...     ...        ...   \n",
       "29666_29666_REF=C_ALT=A          SNP      L      I  37L>I       1          1   \n",
       "29666_29666_REF=C_ALT=G          SNP      L      V  37L>V       1          1   \n",
       "29667_29667_REF=T_ALT=C          SNP      L      P  37L>P       1          1   \n",
       "29669_29669_REF=A_ALT=G          SNP      T      A  38T>A       1          1   \n",
       "29670_29670_REF=C_ALT=T          SNP      T      I  38T>I      13          7   \n",
       "\n",
       "                         infection_rate  fatality_rate  \\\n",
       "descriptor                                               \n",
       "269_269_REF=G_ALT=A          447.212807      13.747245   \n",
       "270_270_REF=A_ALT=G            6.136787       5.394306   \n",
       "271_271_REF=G_ALT=T          447.212807      13.747245   \n",
       "273_273_REF=G_ALT=A         1235.078018       3.558906   \n",
       "274_274_REF=C_ALT=A         1235.078018       3.558906   \n",
       "...                                 ...            ...   \n",
       "29666_29666_REF=C_ALT=A      447.212807      13.747245   \n",
       "29666_29666_REF=C_ALT=G      447.212807      13.747245   \n",
       "29667_29667_REF=T_ALT=C     1235.078018       3.558906   \n",
       "29669_29669_REF=A_ALT=G       22.892683       3.441471   \n",
       "29670_29670_REF=C_ALT=T      590.213587       6.685468   \n",
       "\n",
       "                                                         counted_countries  \n",
       "descriptor                                                                  \n",
       "269_269_REF=G_ALT=A                                  {'united kingdom': 1}  \n",
       "270_270_REF=A_ALT=G                                           {'china': 1}  \n",
       "271_271_REF=G_ALT=T                                  {'united kingdom': 3}  \n",
       "273_273_REF=G_ALT=A                                   {'united states': 1}  \n",
       "274_274_REF=C_ALT=A                                   {'united states': 2}  \n",
       "...                                                                    ...  \n",
       "29666_29666_REF=C_ALT=A                              {'united kingdom': 1}  \n",
       "29666_29666_REF=C_ALT=G                              {'united kingdom': 1}  \n",
       "29667_29667_REF=T_ALT=C                               {'united states': 1}  \n",
       "29669_29669_REF=A_ALT=G                                       {'japan': 1}  \n",
       "29670_29670_REF=C_ALT=T  {'china': 1, 'australia': 1, 'united kingdom':...  \n",
       "\n",
       "[13244 rows x 14 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_muts[unique_muts['VEP']=='missense_variant']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:py3.8_misc]",
   "language": "python",
   "name": "conda-env-py3.8_misc-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
